Add comptime.breakpoint

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102758

This sets a pdb breakpoint to fire whenever we *compile* this
Python code in Dynamo.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

ezyang committed:
Add comptime.breakpoint

This sets a pdb breakpoint to fire whenever we *compile* this
Python code in Dynamo.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add comptime.breakpoint"


This sets a pdb breakpoint to fire whenever we *compile* this
Python code in Dynamo.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/comptime.py:
```diff
@@ -297,6 +301,14 @@ def print_guards():
     comptime(lambda ctx: ctx.print_guards())
 
 
+def breakpoint():
+    def inner(inner_ctx):
+        ctx = inner_ctx.parent()
+        builtins.breakpoint()
+
+    comptime(inner)
```
Noob question, but how do you actually use this API? @comptime.breakpoint on a function?

(nit: could use a quick example in a docstring, so other ppl can figure out how to use it)

ezyang replied:
Just comptime.breakpoint(), just like you do a normal breakpoint

ezyang committed:
Update on "Add comptime.breakpoint"


This sets a pdb breakpoint to fire whenever we *compile* this
Python code in Dynamo.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
